src/v/serde/serde.h

```diff
@@ -480,6 +491,24 @@ void read_nested(iobuf_parser& in, T& t, std::size_t const bytes_left_limit) {
             t.push_back(read_nested<value_type>(in, bytes_left_limit));
         }
         t.shrink_to_fit();
+    } else if constexpr (reflection::is_tristate_v<T>) {
+        int8_t flag = read_nested<int8_t>(in, bytes_left_limit);
+        if (flag == -1) {
+            // disabled
+            t = T{};
+        } else if (flag == 0) {
+            // empty
+            t = T(std::nullopt);
+        } else if (flag == 1) {
+            t = T(read_nested<typename T::value_type>(in, bytes_left_limit));
+        } else {
+            throw serde_exception(fmt_with_ctx(
+              ssx::sformat,
+              "reading type {} of size {}: {} bytes left",
+              type_str<Type>(),
+              sizeof(Type),
+              in.bytes_left()));
+        }
```

Comment:
this exception would be more useful if it also provided context about why it is being thrown: namely, the value of `flag`.